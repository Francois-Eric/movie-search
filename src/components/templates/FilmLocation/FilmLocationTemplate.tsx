import { useEffect, useRef, useState } from "react";
import { getFilmingLocations } from "../../../api/FilmLocation";
import { FilterBar } from "../../molecules/FilterBar/FilterBar";
import { FilmLocationTable } from "../../molecules/FilmLocationTable/FilmLocationTable";
import { LoadingProgress } from "../../atom/LoadingProgress/LoadingProgress";
import {
  FilmLocationFilters,
  FilmLocationNormalizedField,
} from "../../../types";
import { useInfiniteScroll } from "../../../hooks/useInfiniteScroll";

export const FilmLocationTemplate = () => {
  const tableContainerRef = useRef<HTMLDivElement>(null);
  const [count, setCount] = useState<number>(0);
  const [filmingLocations, setFilmingLocations] = useState<
    FilmLocationNormalizedField[]
  >([]);
  const [filters, setFilters] = useState<FilmLocationFilters[]>([]);
  const [selectedFilters, setSelectedFilters] = useState<{
    [key: string]: string[];
  }>({});
  const [hasFailed, setHasFailed] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [offset, setOffset] = useState(0);

  const { resetScroll } = useInfiniteScroll(
    tableContainerRef,
    () => {
      if (offset < count) {
        setOffset(filmingLocations.length);
      }
    },
    50
  );

  useEffect(() => {
    setOffset(0);
    resetScroll();
    setFilmingLocations([]);
    const getData = async () => {
      setIsLoading(true);
      try {
        const { records, filters, count } = await getFilmingLocations(
          0,
          selectedFilters
        );
        setFilmingLocations(records);
        setFilters(filters);
        setCount(count);
      } catch (error) {
        setHasFailed(true);
      } finally {
        setIsLoading(false);
      }
    };
    getData();
  }, [selectedFilters]);

  useEffect(() => {
    const getData = async () => {
      setIsLoading(true);
      try {
        const { records } = await getFilmingLocations(offset, selectedFilters);
        setFilmingLocations([...filmingLocations, ...records]);
      } catch (error) {
        setHasFailed(true);
      } finally {
        setIsLoading(false);
      }
    };
    getData();
  }, [offset]);

  if (hasFailed) {
    <p>Something went wrong, try reloading page</p>;
  }

  return (
    <div>
      {filters?.length > 0 && (
        <FilterBar
          filters={filters}
          selectedFilters={selectedFilters}
          setSelectedFilters={setSelectedFilters}
        />
      )}
      <LoadingProgress count={filmingLocations.length} total={count} />
      <FilmLocationTable
        records={filmingLocations}
        tableContainerRef={tableContainerRef}
        isLoading={isLoading}
      />
    </div>
  );
};
