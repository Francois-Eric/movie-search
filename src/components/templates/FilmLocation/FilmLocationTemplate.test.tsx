import { act, render, waitFor } from "@testing-library/react";
import {
  mockedFilmLocationFilters,
  mockedFilmLocationRecords,
} from "../../../mocks/FilmLocationMocks";
import { FilmLocationTemplate } from "./FilmLocationTemplate";

jest.mock("../../molecules/FilterBar/FilterBar", () => {
  return {
    FilterBar: () => {
      return <div>Mocked FilterBar</div>;
    },
  };
});

jest.mock("../../../hooks/useInfiniteScroll", () => ({
  useInfiniteScroll: () => ({
    resetScroll: () => {},
  }),
}));

jest.mock("../../../api/FilmLocation", () => ({
  getFilmingLocations: () => ({
    records: mockedFilmLocationRecords,
    filters: mockedFilmLocationFilters,
    count: mockedFilmLocationRecords.length,
  }),
}));

describe("<FilmLocationTemplate/>", () => {
  test("should render correctly", async () => {
    let component: ReturnType<typeof render>;
    await act(async () => {
      component = await render(<FilmLocationTemplate />);
    });

    await waitFor(() => {
      expect(
        component.getByText(mockedFilmLocationRecords[0].nom_producteur)
      ).toBeTruthy();
      expect(component.container).toMatchSnapshot();
    });
  });
});
