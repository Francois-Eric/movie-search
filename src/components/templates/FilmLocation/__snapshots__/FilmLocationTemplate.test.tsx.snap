// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilmLocationTemplate/> should render correctly 1`] = `
<div>
  <div>
    <div>
      Mocked FilterBar
    </div>
    <div
      class="LoadingProgress"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2 css-1sr7afs-MuiTypography-root"
      >
        2 résultats chargés sur 2
      </h6>
      <span
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="100"
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-eglki6-MuiLinearProgress-root"
        role="progressbar"
      >
        <span
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-5xe99f-MuiLinearProgress-bar1"
          style="transform: translateX(0%);"
        />
      </span>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1fw5qro-MuiPaper-root"
    >
      <div
        class="MuiTableContainer-root FilmLocationTable__container css-rorn0c-MuiTableContainer-root"
      >
        <table
          class="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
        >
          <thead
            class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                scope="col"
              >
                Année
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                scope="col"
              >
                Type
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                scope="col"
              >
                Nom
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                scope="col"
              >
                Réalisateur
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                scope="col"
              >
                Producteur
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                scope="col"
              >
                Adresse
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                scope="col"
              >
                Code Postal
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                annee_tournage_1
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                type_tournage_1
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                nom_tournage_1
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                nom_realisateur_1
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                nom_producteur_1
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                adresse_lieu_1
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                ardt_lieu_1
              </td>
            </tr>
            <tr
              class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                annee_tournage_2
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                type_tournage_2
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                nom_tournage_2
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                nom_realisateur_2
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                nom_producteur_2
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                adresse_lieu_2
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                ardt_lieu_2
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
